File: src/java/magma/app/rule/PrefixRule.java
package magma.app.rule;

import magma.api.optional.OptionalLike;
import magma.api.optional.Optionals;
import magma.app.node.Node;

public record PrefixRule(String prefix, Rule rule) implements Rule {
    @Override
    public OptionalLike<Node> lex(final String input) {
        if (!input.startsWith(this.prefix))
            return Optionals.empty();

        final var prefixLength = this.prefix.length();
        final var slice = input.substring(prefixLength);
        return this.rule.lex(slice);
    }

    @Override
    public OptionalLike<String> generate(final Node node) {
        return this.rule.generate(node)
                .map(result -> this.prefix + result);
    }
}

File: src/node/magma/rule/PrefixRule.ts
/*import magma.CompileError;*/
/*import magma.node.Node;*/
/*import magma.result.Err;*/
/*import magma.result.Ok;*/
/*import magma.result.Result;*/
/*import java.util.Optional;*/
/*public record PrefixRule(String prefix, Rule rule) implements Rule {
	private Optional<Node> lex0(final String input) {
		if (input.startsWith(this.prefix)) return this.rule.lex(input.substring(this.prefix.length())).findValue();
		return Optional.empty();
	}

	private Optional<String> generate0(final Node node) {
		return this.rule.generate(node).findValue().map(result -> this.prefix + result);
	}

	@Override
	public Result<Node, CompileError> lex(final String input) {
		return this.lex0(input).<Result<Node, CompileError>>map(Ok::new).orElseGet(() -> new Err<>(new CompileError()));
	}

	@Override
	public Result<String, CompileError> generate(final Node node) {
		return this.generate0(node).<Result<String, CompileError>>map(Ok::new).orElseGet(() -> new Err<>(new CompileError()));
	}
}*/
